### PR TITLE
Edit tags directly instead of using CommandModifier in BasicEngine.allocate_qubit

### PR DESCRIPTION
**Pull request by Strilanc**
If the allocate command could somehow re-enter then this is not a no-op change, but there's no test checking for that behavior so I assume it doesn't happen?

### CODE REVIEW

**Comment by damiansteiger** [comment]
> If the allocate command could somehow re-enter then this is not a no-op change, but there's no test checking for that behavior so I assume it doesn't happen?

Could you explain this in more detail?

File: projectq/cengines/_basics.py [hunk 1]
```diff
@@ -107,26 +107,13 @@ def allocate_qubit(self, dirty=False):
         """
         new_id = self.main_engine.get_new_qubit_id()
         qb = Qureg([Qubit(self, new_id)])
+        cmd = Command(self, Allocate, (qb,))
         if dirty:
             from projectq.meta import DirtyQubitTag
-            if self.is_meta_tag_supported(DirtyQubitTag):
```
Comment (damiansteiger) [inline projectq/cengines/_basics.py]: Why is this condition deleted?

**Comment by Strilanc** [comment]
> Could you explain this in more detail?

If an engine further down the chain feeds a Not gate back into the top of the chain, we may end up back at this line of code while the command modifier is still is in effect. So the dirty tag would be added to the Not gate.

This PR changes that behavior, and I was pointing that out. I suspect it's probably a bug fix instead of a breaking change, but it's worth keeping in mind.

File: projectq/cengines/_basics.py (hunk 1 above)
Comment (Strilanc) [inline projectq/cengines/_basics.py]: Oops. Restored.

**Comment by damiansteiger** [comment]
Thanks for the change, the code is now cleaner 👍

**Pull request merged**